File: NeetCodeIO/981TimeBasedKey-ValueStore/TimeBasedKey-ValueStore.cpp
class TimeMap {
public:
    map<string,vector<pair<string,int>>> m;
    TimeMap() {
        
    }
    
    void set(string key, string value, int timestamp) {
        // Map of maps -> Time:O(1), Space:O(1)
        if (m.find(key) == m.end()) {
            vector<pair<string,int>> tempVec;
            m[key] = tempVec;
        }
        pair<string,int> tempPair(value,timestamp);
        m[key].push_back(tempPair);
    }
    
    string get(string key, int timestamp) {
        // Binary Search -> Time:O(log(n)), Space:O(1) n:values.size()
        string ans = "";
        if (m.find(key) == m.end()) {
            return ans;
        }
        vector<pair<string,int>> values = m[key];
        int mid, start = 0, end = values.size()-1;
        while (start <= end) {
            mid = (start + end) / 2;
            if (values[mid].second == timestamp) {
                return values[mid].first;
            }
            else if (values[mid].second < timestamp) {
                ans = values[mid].first;
                start = mid + 1;
            }
            else {
                end = mid - 1;
            }
        }
        return ans;
    }
};

File: NeetCodeIO/981TimeBasedKey-ValueStore/TimeBasedKey-ValueStore.py
class TimeMap:

    def __init__(self):
        self.d = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        # Dict of Dict -> Time:O(1), Space:O(1)
        if key not in self.d:
            self.d[key] = []
        self.d[key].append([timestamp, value])
        
    def get(self, key: str, timestamp: int) -> str:
        # Binary Search -> Time:O(log(n)), Space:O(1) n:len(values)
        if key not in self.d:
            return ""
        ans = ""
        values = self.d[key]
        start, end = 0, len(values) - 1
        while start <= end:
            mid = (start+end) // 2
            if values[mid][0] == timestamp:
                return values[mid][1]
            elif values[mid][0] < timestamp:
                ans = values[mid][1]
                start = mid + 1
            else:
                end = mid - 1
        return ans

# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)
